src/tools/rbd/action/Import.cc

```diff
@@ -173,9 +174,31 @@ static int do_import(librbd::RBD &rbd, librados::IoCtx& io_ctx,
 
     // write as much as we got; perhaps less than imgblklen
     // but skip writing zeros to create sparse images
-    if (!bl.is_zero()) {
-      C_Import *ctx = new C_Import(*throttle, image, bl, image_pos);
-      ctx->send();
+    size_t b_offset = 0;
+    size_t b_length = 0;
+    size_t b_end = 0;
+    size_t p_offset = 0;
+    size_t p_shift = 0;
+    while (b_end < blklen) {
+      if (b_end + pagesize > blklen)
+        p_shift = blklen - b_end;
+      else
+        p_shift = pagesize;
+
+      bufferptr ptr(blk_ptr, p_offset, p_shift);
+      p_offset += p_shift;
+      b_end += p_shift;
+      if (!ptr.is_zero())
+        b_length += p_shift;
+      if ((ptr.is_zero() && b_length != 0) || b_end == blklen) {
```

Comment:
Nit: cache result of ```is_zero``` (since it isn't constant time op) or restructure the if blocks.